src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -37,6 +37,37 @@ internal class DiagnosticComputer
         /// pressure when there are large number of open documents across different projects to be analyzed by background analysis.
         /// </summary>
         private static CompilationWithAnalyzersCacheEntry? s_compilationWithAnalyzersCache = null;
+
+        /// <summary>
+        /// Set of high priority diagnostic computation tasks which are currently executing.
+        /// Any new high priority diagnostic request is added to this set before the core diagnostics
+        /// compute call is performed, and removed from this list after the computation finishes.
+        /// Any new normal priority diagnostic request first waits for all the high priority tasks in this set
+        /// to complete, and moves ahead only after this list becomes empty.
+        /// </summary>
+        /// <remarks>
+        /// Read/write access to this set is guarded by <see cref="s_gate"/>.
+        /// </remarks>
+        private static ImmutableHashSet<Task> s_highPriorityComputeTasks = ImmutableHashSet<Task>.Empty;
+
+        /// <summary>
+        /// Cancellation token source for normal priority diagnostic computation tasks which are currently executing.
+        /// Any new high priority diagnostic request fires cancellation on this cancellation token source
+        /// to avoid resource contention between normal and high priority requests. It also initializes this field
+        /// with a new cancellation token source for subsequent requests.
+        /// Canceled normal priority diagnostic requests are re-attempted from scratch after all the high priority requests complete.
+        /// </summary>
+        /// <remarks>
+        /// Read/write access to this field is guarded by <see cref="s_gate"/>.
+        /// </remarks>
+        private static ImmutableHashSet<CancellationTokenSource> s_normalPriorityCancellationTokenSources = ImmutableHashSet<CancellationTokenSource>.Empty;
```

Comment:
note: i used a set again.  you can def just have one CTS (that you overwrite) if you want.  But with either approach, you still only need to create the linked CTS once afaict.  I'll point that out below.